### PR TITLE
[PyTorch] Integration test for Megatron-LM

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

https://github.com/NVIDIA/TransformerEngine/pull/1033 broke Megatron-LM's wrappers for the LayerNorm and RMSNorm modules:
- We renamed the `hidden_size` arg to `normalized_shape` in order to match [`torch.nn.LayerNorm`](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html), but Megatron-LM treats `hidden_size` as a kwarg:
https://github.com/NVIDIA/Megatron-LM/blob/aded519cfb1de2abf96f36ca059f992294b7876f/megatron/core/extensions/transformer_engine.py#L65.
This PR adds logic to handle the `hidden_size` arg and print a deprecation warning.
- Megatron-LM has the option to initialize modules in CPU, but TE's LayerNorm and RMSNorm ops assume they are initialized on GPU. This PR allows LayerNorm and RMSNorm ops to be initialized on CPU, although it still assumes it is transferred to GPU before performing compute.

To help detect these issues in the future, I've also added an integration test that runs Megatron-LM to train a very small GPT model.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor
- [x] Testing

## Changes

- Handle deprecated `hidden_size` arg in LayerNorm and RMSNorm modules
- Allow LayerNorm and RMSNorm operations to be initialized on CPU
- Add Megatron-LM integration test

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by timmoon10** [comment]
Pipeline 20338114

**Comment by timmoon10** [comment]
Pipeline 20444324

File: transformer_engine/pytorch/ops/fuser.py [hunk 1]
```diff
@@ -135,7 +135,11 @@ def forward(
                     requires_grad = any(any(x.requires_grad for x in xs) for xs in extra_inputs)
             for idx in basic_op_idxs:
                 basic_op_ctxs[idx].requires_grad = requires_grad
-            x.requires_grad_(requires_grad=requires_grad)
+            if requires_grad != x.requires_grad:
+                if requires_grad:
+                    x.requires_grad_()
+                else:
+                    x = x.detach()
```
Comment (timmoon10) [inline transformer_engine/pytorch/ops/fuser.py:142]: This fixes a te.Sequential bug that was exposed by Mcore. When running in eval mode, we want `x.requires_grad=False` so that the op knows that it doesn't need to prepare for that grad. However, PyTorch sometimes complains if you change a tensor's `requires_grad` from `True` to `False` (i.e. when the tensor is not a leaf in the autograd graph). Detaching the tensor works around this case.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 2]
```diff
@@ -65,15 +65,28 @@ class RMSNorm(_RMSNormOp):
...
         eps: float = 1e-5,
         sequence_parallel: Optional[bool] = None,  # legacy
         params_dtype: Optional[torch.dtype] = None,  # deprecated
         zero_centered_gamma: bool = False,
+        hidden_size: Optional[int] = None,  # deprecated
         **kwargs,
     ) -> None:
 
         # Handle deprecated options
+        if normalized_shape is None:
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/module/rmsnorm.py:78]: what's the reason to not check for `hidden_size == None` here?

File: transformer_engine/pytorch/ops/basic/layer_norm.py [hunk 3]
```diff
@@ -184,17 +185,21 @@ def op_forward(
     ) -> torch.Tensor:
 
         # Check tensor dims
+        weight = self.weight
+        weight_dims = tuple(weight.size())
         input_dims = tuple(input_.size())
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/ops/basic/layer_norm.py:190]: apparently `torch.Size` is a subclass of `tuple` so tuple creation probably not needed

File: transformer_engine/pytorch/ops/basic/rmsnorm.py [hunk 4]
```diff
@@ -165,17 +168,21 @@ def op_forward(
     ) -> torch.Tensor:
 
         # Check tensor dims
+        weight = self.weight
+        weight_dims = tuple(weight.size())
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/ops/basic/rmsnorm.py:172]: no need to `tuple`ize

**Review by sudhakarsingh27 — APPROVED** [review]
